design/external-authorization-design.md

```diff
@@ -0,0 +1,669 @@
+const SupportProtocolVersion2 SupportProtocolVersion = "v2"
+
+type SupportService struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   SupportServiceSpec   `json:"spec,omitempty"`
+	Status SupportServiceStatus `json:"status,omitempty"`
+}
+
+type SupportServiceStatus struct {
+	Conditions []Condition `json:"conditions"`
+}
+
+type SupportServiceSpec struct {
+	// +required
+	Service contourv1.Service `json:"service"`
+
+	// The load balancing policy for sending requests to the service.
+	// +optional
+	LoadBalancerPolicy *contourv1.LoadBalancerPolicy `json:"loadBalancerPolicy,omitempty"`
+
+	// This field sets the version of the GRPC protocol that Envoy uses to
+	// send requests to the support service. Since Contour always uses the
+	// v2 Envoy API, this is currently fixed at "v2". However, other
+	// protocol options will be available in future.
+	//
+	// +kubebuilder:validation:Enum=v2
+	ProtocolVersion SupportProtocolVersion `json:"protocolVersion,omitempty"`
+}
+```
+
+The `Conditions` field in `SupportServicesStatus` follows the standardized Contour structure (proposed in #2642).
+This type should be shared with other Contour API status types.
+
+This document does not yet define how Contour monitors the support service.
+However, the presence of the `Conditions` field allows Contour to expose the status of the support service fairly directly.
+At minimum, Contour should monitor the underlying Kubernetes service and expose whether there are sufficient healthy endpoints.
+
+Each `SupportService` CRD generates a unique upstream Envoy [Cluster][4], which will emit standard Envoy metrics.
+Note that the Envoy cluster name can be non-obvious, so exposing it in status may be helpful.
+
+If the `Service` refers to a Kubernetes `ExternalName`, Contour should program Envoy to send the traffic to the external destination.
+
+The `SupportService` CRD re-uses the `Service` type from the `projectcontour.io/v1` API.
+However, the setting following fields can generate a validation errors:
+
+- `Mirror` must not be set
+- `Weight` must not be set
+- `Protocol` may only be set to `h2`
+
+Note that the `Service` type does not include a field for a namespace name.
+This constrains `SupportService` resources to be in the same namespace as the `Service` they expose,
+ensuring that whoever creates the `SupportService` also has authority over the `Service` backing it.
+
+### HTTPProxy Changes
+
+```Go
+// SupportServiceReference names a SupportService resource.
+type SupportServiceReference struct {
+	// API version of the referent.
+	// +required
+	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,5,opt,name=apiVersion"`
+	// Kind of the referent.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	// +required
+	Kind string `json:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`
```

Comment:
Should we make this optional with the behavior being that empty is equivalent to `SupportService`? I know you're basing this off `TypedLocalObjectReference`, but I think it's worth considering a slightly different behavior in this case.

Reply:
If we default the kind, then users would do:
```yaml
    authorization:
      serviceRef:
        apiVersion: projectcontour.io/v1alpha1
        name: htpasswd
        namespace: projectcontour-auth
```

Maybe we could remove or default `apiVersion` as well, depending on how we manage the versioning of this type. I'm also open to just removing these fields altogether.

Reply:
I think it's better to default them than remove them, that gives us the chance to use other kinds later.

Reply:
Given that this typed is named `SupportServiceReference`, the `Kind` field seems like it'd be either redundant or contradictory, unless the intent is to allow specializations of SupportService (e.g. AuthSupportService, RateLimitingSupportService)?